### ISSUE TITLE
union: doesn't support --fast-list which slows down vfs/refresh

### ISSUE THREAD

**Issue by rsierra** [rclone/rclone#3000]
#### What is the problem you are having with rclone?

Doing a vfs/refresh remote command over a union with a remote take much time than over the remote.

I tried with the code but it's too much golang for my level 😅. Could it be that it is not using the `-fast-list` option with the remote through the union mount point? Maybe it's just a not covered case.

I notice this because I was getting timeouts with large folders if I mount the union instead the remote.

#### What is your rclone version (output from `rclone version`)

```
rclone v1.46
- os/arch: linux/amd64
- go version: go1.11.5
```

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Linux 4.20.8-arch1-1-ARCH x86_64

####  Which cloud storage system are you using? (eg Google Drive)

Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

With the config:

```
[gdrive]
type = drive
client_id = ...
client_secret = ...
token = ...

[union]
type = union
remotes = /srv/data gdrive:/data
```

And the remote mounted with remote control:

```
rclone mount union: /mnt/data-union \
   --allow-other --allow-non-empty \
   --buffer-size 512M \
   --dir-cache-time 48h \
   --fast-list \
   --use-mmap \
   --read-only \
   --rc \
   --vfs-read-chunk-size 8M \
   --vfs-read-chunk-size-limit 256M
```

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

I put just the response times because the action it's going well in any case.

`time rclone rc vfs/refresh recursive=true dir=/path`

```
{
	"result": {
		"/path": "OK"
	}
}

real	0m30.601s
user	0m0.008s
sys	0m0.014s
```

But if I mount the `gdrive:/data` instead the `union`:

`time rclone rc vfs/refresh recursive=true dir=/path`

```
{
	"result": {
		"/path": "OK"
	}
}

real	0m2.438s
user	0m0.007s
sys	0m0.011s
```

Thank you for the time in advance and sorry if something is wrong with the report.

**Comment by RXWatcher1** [comment]
I am also seeing this behavior and I was about to log an issue over it. I thought it never completed. It looks like it does but takes a very long time.

**Comment by calisro** [comment]
I wonder if you could add 

fast_list = true

to the config for the underlying remote in the rclone.conf and see a time difference as then it should use fast-list with the underlying remote that's part of the union.

https://github.com/ncw/rclone/issues/2697

**Comment by RXWatcher1** [comment]
> I wonder if you could add
> 
> fast_list = true
> 
> to the config for the underlying remote in the rclone.conf and see a time difference as then it should use fast-list with the underlying remote that's part of the union.
> 
> #2697

In my use case I'm trying to union multiple Team Drives and I believe fast-list isn't recommended for Team Drives. The main advantage for me to use rclone union is I want a single systemd unit to manage it all vs mounting indv Team Drives and using something like Mergerfs to union them.

**Comment by calisro** [comment]
I think that depends on what you're doing against the team drives because of the delay in getting consistent reads.

"Warning: Rclone has detected that you’re using --fast-list with sync/copy/move to or from a Google Team Drive remote. Due to the way Team Drives currently cache the full directory listing there can be situations where Rclone is unable to see recently created files on the drive and cause the command to miss or re-copy files. Running the command without --fast-list should allow Rclone to properly detect all files on the remote. Waiting a currently unknown amount of time after a file has been uploaded (Estimated to be less than an hour) should also be effective."

**Comment by RXWatcher1** [comment]
I added `fast_list = true` to the Team Drive remotes and tried it. It seemed promising on team drives with little content. It took 5 mins for the refresh to complete on the small team drives however when I attempted it on Team Drives with more content it took over an hour. I can't tell you the exact time because I had the timeout set for 60 minutes and it timed out at that point.

**Comment by rsierra** [comment]
> I wonder if you could add
> 
> fast_list = true
> 
> to the config for the underlying remote in the rclone.conf and see a time difference as then it should use fast-list with the underlying remote that's part of the union.
> 
> #661

I've got the same response times with `fast_list = true` in the config file 😢 .

**Comment by ncw** [comment]
> Could it be that it is not using the `--fast-list` option with the remote through the union mount point? Maybe it's just a not covered case

That is exactly the problem... The Union mount doesn't support --fast-list.

**Comment by ncw** [comment]
@Max-Sum what do you think about this issue as the next thing to work on for the union backend? It would be super useful for people using `vfs/refresh` with the recursive flag.

**Comment by ncw** [comment]
@Max-Sum has just implemented this and I've merged it to [the latest beta](https://beta.rclone.org).

Please give it a go :-)

**Comment by Max-Sum** [comment]
Since both local and gdrive is used, I don't think my implementation could work though...
@ncw Do you think iterating for upstreams without ListR supported is fine?

**Comment by Visorask** [comment]
So it seems to work fine without local: Took about 10 minutes

`2020/05/10 22:25:47 DEBUG : rclone: Version "v1.51.0-300-g54b16bd0-beta" starting with parameters ["rclone" "rc" "vfs/refresh" "recursive=true" "-vvv" "--timeout" "60m"]
{
	"result": {
		"": "OK"
	}
}
2020/05/10 22:35:26 DEBUG : 4 go routines active`

With local:
`rclone rc vfs/refresh recursive=true -vvv --timeout 60m
2020/05/10 22:37:54 DEBUG : rclone: Version "v1.51.0-300-g54b16bd0-beta" starting with parameters ["rclone" "rc" "vfs/refresh" "recursive=true" "-vvv" "--timeout" "60m"]`

It's still running and I've had it running for about 20m now. So your theory on it not working with local is correct.

@Max-Sum

**Comment by rsierra** [comment]
First of all, thx for the effort :-).

I have also repeated a similar test again and the result is the same:

- Only with remote:

```bash
$ time rclone-v1.51.0-301-g10a6a92e-beta-linux-amd64/rclone rc vfs/refresh recursive=true dir=/test
{
        "result": {
                "/test": "OK"
        }
}

real    0m3.679s
user    0m0.027s
sys     0m0.010s
```

- Only with union (local+remote):

```bash
$ time rclone-v1.51.0-301-g10a6a92e-beta-linux-amd64/rclone rcvfs/refresh recursive=true dir=/test
{
        "result": {
                "/test": "OK"
        }
}

real    0m40.209s
user    0m0.019s
sys     0m0.013s
```

**Comment by Max-Sum** [comment]
@rsierra Yes, we face a dilemma here. If we enable fast list only when all upstreams support fast list, it would be a pity for local+remote case. However, if we enable it when one of the upstreams support, it would create unnecessary requests to the unsupported upstream and possibly damage the performance.

I could make a compromise, enable fast list if all upstreams are in one of these cases:
 - fast list supported
 - local
 - ram

**Comment by ncw** [comment]
> @rsierra Yes, we face a dilemma here. If we enable fast list only when all upstreams support fast list, it would be a pity for local+remote case. However, if we enable it when one of the upstreams support, it would create unnecessary requests to the unsupported upstream and possibly damage the performance.
> 
> I could make a compromise, enable fast list if all upstreams are in one of these cases:
> 
> * fast list supported
> * local
> * ram
>   What do you think @ncw?

That would work OK. You could use the `IsLocal` flag on `Fs.Features()` to do that detection - that is what it is for.

Note that you can use `fs/walk.ListR` to turn `Fs.List()` into `Fs.ListR()`.

https://github.com/rclone/rclone/blob/bd4b91bd57c04e95c5a25fd0ba2d41d32faf5ab8/fs/walk/walk.go#L140

The other alternative would be to implement ListR for the local backend. This would probably be slightly more efficient than using `walk.ListR` on it and might be useful elsewhere

**Comment by Max-Sum** [comment]
Thank you for your notice.
Would implementing ListR on local affect the performance for non-union users?

**Comment by Max-Sum** [comment]
@ncw What about memory backend. I think turning on `IsLocal` for memory is ok?

**Comment by ncw** [comment]
> Wouldn't implementing ListR on local affect the performance for non-union users?

Done right it would increase performance... But it is probably hard to do better than the rclone version which uses multithreaded lists.

> What about memory backend. I think turning on IsLocal for memory is ok?

That is OK I think - I thought of doing it a while back!

**Comment by ncw** [comment]
I've merged @Max-Sum 's fix to enable --fast-list when local backend is part of the union to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52

**Comment by rsierra** [comment]
Great news! Thank you all!
